lib/bot_framework/session.rb

```diff
@@ -0,0 +1,422 @@
+              @sending_batch = false
+              start_batch if @batch_started # What is this :o
+              yield(err) if block_given?
+            end
+          else
+            @sending_batch = false
+            start_batch if @batch_started
+            yield(err) if block_given?
+          end
+        end
+      end
+    end
+
+    
+    # Begin a new dialog
+    def begin_dialog(id, args = nil)
+      logger.info "Beginning new dialog #{id}"
+      id = resolve_dialog_id(id)
+      dialog = find_dialog(id)
+      raise "Dialog #{id} Not found" unless dialog
+      push_dialog(id: id, state: {})
+      start_batch
+      dialog.begin(self, args)
+      self
+    end
+
+    def replace_dialog(id, args = nil)
+      logger.info "Session replace dialog #{id}"
+      id = resolve_dialog_id(id)
+      dialog = find_dialog(id)
+      raise "Dialog #{id} Not found" unless dialog
+
+      # Update the stack and start dialog
+      pop_dialog
+      push_dialog(id: id, state: {})
+      start_batch
+      dialog.begin(self, args)
+      self
+    end
+
+    # End conversation with the user
+    def end_conversation(message = nil, _args = {})
+      if message
+        # TODO: sent message
+      end
+      # Clear private conversation data
+      @private_conversation_data = {}
+
+      # Clear stack and save
+      logger.info 'End conversation'
+      ss = @session_state
+      ss.call_stack = []
+      send_batch
+      self
+    end
+
+    def end_dialog(message = nil, args = {})
+      if message
+        # TODO: end_dialog_with_result
+      end
+      cur = cur_dialog
+      if cur
+        if message
+          m = if (message.is_a? String) || (message.is_a? Array)
+                create_message(cur_library_name, message, args)
+              else
+                message
+              end
+          @msg_sent = true
+          prepare_message(m)
+          @batch.push(m)
+        end
+
+        # Pop the dialog of the stack and resume the parent
+        logger.info 'Session.end_dialog'
+        child_id = cur.id
+        cur = pop_dialog
+        start_batch
+        if cur
+          dialog = find_dialog(cur.id)
+          if dialog
+            dialog.dialog_resumed(self, resumed: :completed, response: true, child_id: child_id)
+          else
+            # Bad dialog !! , we should never reach here
+            raise "Can't resume , missing parent"
+          end
+        end
+      end
+    end
+
+    # Ends current dialog and returns a value to the caller
+    def end_dialog_with_result(result)
+      # Validate call stack
+      cur = cur_dialog
+      if cur
+        # Validate the result
+        result ||= {}
+        result[:resumed] ||= :completed
+        result.child_id = cur.id
+        logger.info 'Session.end_dialog_with_result'
+
+        # Pop dialog of the stack and resume parent dialog
+        cur = pop_dialog
+        start_batch
+        if cur
+          dialog = find_dialog(cur.id)
+          if dialog
```

Comment:
Use a guard clause instead of wrapping the code inside a conditional expression.